liwords-ui/src/chat/players.tsx

```diff
@@ -0,0 +1,294 @@
+            <Player
+              sendMessage={sendMessage}
+              className={className}
+              key={p.uuid}
+              {...p}
+            />
+          ))}
+        </>
+      );
+    },
+    [sendMessage]
+  );
+
+  const filterPlayerListBySearch = useCallback(
+    (searchTerm: string, list: Partial<FriendUser>[]) => {
+      if (searchTerm?.length) {
+        return list.filter((u) =>
+          u.username?.toLowerCase().startsWith(searchTerm.toLowerCase())
+        );
+      } else {
+        return list;
+      }
+    },
+    []
+  );
+
+  const transformAndFilterPresences = useCallback(
+    (
+      presenceUsers: PresenceEntity[],
+      searchTerm: string
+    ): Partial<FriendUser>[] => {
+      const presencePlayers = presenceUsers
+        .filter((p) => !p.anon)
+        .sort((a, b) => {
+          if (a.username < b.username) {
+            return -1;
+          }
+          if (a.username > b.username) {
+            return 1;
+          }
+          return 0;
+        })
+        .map(
+          (p): Partial<FriendUser> => {
+            return {
+              username: p.username,
+              uuid: p.uuid,
+              channel: new Array<string>().concat(p.channel),
+            };
+          }
+        )
+        .filter((u) => u.username?.toLowerCase() !== username.toLowerCase());
+      return searchTerm?.length
+        ? presencePlayers.filter((u) =>
+            u.username?.toLowerCase().startsWith(searchTerm.toLowerCase())
+          )
+        : presencePlayers;
+    },
+    [username]
+  );
+
+  useEffect(() => {
+    window.addEventListener('resize', setHeight);
+    return () => {
+      window.removeEventListener('resize', setHeight);
+    };
+  }, [setHeight]);
+
+  const getPresenceLabel = (channelType: string) => {
+    switch (channelType) {
+      case 'lobby':
+        return 'IN LOBBY';
+      case 'game':
+        return 'YOUR OPPONENT';
+      case 'gametv':
+        return 'OBSERVERS';
+    }
+    return 'IN ROOM';
+  };
+  return (
+    <div className="player-list">
+      <Form name="search-players">
+        <Input
+          placeholder="Search players"
+          name="search-players"
+          onChange={handleSearchChange}
+          value={searchText}
+          autoComplete="off"
+        />
+      </Form>
+      <div
+        className={`player-sections ${
+          props.defaultChannelType ? props.defaultChannelType : ''
+        }`}
+        style={
+          maxHeight
+            ? {
+                maxHeight: maxHeight,
+                overflowY: 'auto',
+              }
+            : undefined
+        }
+      >
+        <section className="friends">
+          {loggedIn && <div className="breadcrumb">FRIENDS</div>}
+          {loggedIn &&
+            renderPlayerList(
+              filterPlayerListBySearch(searchText, Object.values(friends)),
+              'friends'
+            )}
+          {loggedIn && Object.values(friends).length === 0 && (
```

Comment:
(performance) each render calls `Object.values(friends)` several times and each call allocates a new copy. maybe `useMemo` it if we really need the values in friends values?
```js
const friendsValues = useMemo(() => Object.values(friends), [friends]);
```